README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# icerui: Calculate Uncertainty Intervals for Incremental Cost-Effectiveness Ratios (ICERs)

<!-- badges: start -->

<!-- badges: end -->

This package provides several methods for calculating uncertainty
intervals (confidence intervals, credible intervals) for the incremental
cost-effectiveness ratio (ICER), which is the ratio of incremental costs
to incremental effects. It includes the bootstrap percentile method,
Fieller method and bootstrap acceptability method.

> :warning: **Early stages of development.** This package is in the
> early stages of development. Please do not assume there will be no
> breaking changes and check back regularly.

## Installation

You can install the latest “stable” version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("tristansnowsill/icerui")
```

## Example

The package includes a single example dataset `ex_smoking` constructed
from the [BCEA](http://www.statistica.it/gianluca/software/bcea/)
package, which we will use here:

``` r
library(icerui)

confint(ex_smoking, method = "fieller")
#>                            2.5%      97.5%
#> No treatment           98.26914 4910.70352
#> Self-help              74.72212 -263.56608
#> Individual counselling 37.72638  -82.84373
```

There are three methods currently implemented:

  - `fieller`, Fieller method (parametric)
  - `percentile`, Bootstrap percentile (non-parametric)
  - `acceptability`, Bootstrap acceptability (non-parametric)

The bootstrap percentile method is implemented as suggested in Glick et
al. (2014).

It is ***strongly recommended*** that you check the confidence intervals
produced are sensible.

The `uiplot` function can help with this:

``` r
uiplot(ex_smoking, method = "percentile", comparison = 1, graph = "base")
```

<img src="man/figures/README-uiplot-example-1.png" width="100%" />

## Bibliography

  - Glick HA, Doshi JA, Sonnad SS, Polsky D. (2014) Economic evaluation
    in clinical trials (2nd Edition). Oxford: Oxford University Press.
    ISBN 978-0-19-968502-8. doi: 10.1093/med/9780199685028.001.0001


NEWS.md
# icerui (development version)

The package is being written to use S3 objects behind the scenes. When this is complete you will be able to run code like:

```r
library(icerui)

cea1 <- BCEA::bcea(...)
ui1 <- icerui(cea1)
ui1
confint(ui1)
plot(ui1)

cea2 <- heemod::run_psa(...)
ui2 <- icerui(cea2, method = "percentile")
```

You will be able to specify the comparisons you want confidence intervals for the ICERs for in a number of ways.

The formula notation will be the most flexible:

```r
# ICER of treatment vs. control
icerui(x, comparison = treatment ~ control)

# ICERs of trtC vs. trtA and trtC vs. trtB
icerui(y, comparison = trtC ~ list(trtA, trtB))

# ICERs of trtC vs. trtB, trtC vs. trtA and trtB vs. trtA
# (ignores invalid trtB vs. trtB comparison)
icerui(y, comparison = list(trtC, trtB) ~ list(trtB, trtA))

# Same as above
# (. on either side mirrors other side)
icerui(y, comparison = list(trtC, trtB, trtA) ~ .)
```

For BCEA and heemod PSA objects you can also specify comparisons with character vectors, but the behaviour is dependent on whether the input is a BCEA object or heemod PSA object:
* For BCEA objects, `bcea$ref` will be compared to each option in the character vector
* For heemod PSA objects, each strategy listed will be compared to `psa$model$central_strategy`

For BCEA objects you can also specify comparisons by integers instead of strings.

For BCEA and heemod PSA objects if you do not specify a comparison, it will behave as if you asked for all comparisons (`bcea$ref` vs. each in `bcea$comp` or all strategies versus `psa$model$central_strategy`)

## Other changes

* `uiplot` is deprecated in favour of `plot.icerui`
* `fieller`, `bspercent` and `bsaccept` are now exported for direct use (e.g. if you don't have a BCEA or heemod PSA object)


tests/testthat.R
library(testthat)
library(icerui)

test_check("icerui")


tests/testthat/test-confint-bcea.R
context("Obtaining uncertainty intervals from BCEA::bcea() objects")

test_that("fieller method with indices", {
  # The ex_simple dataset is constructed such that incremental costs and
  # effects are bivariate normally distributed with mean 5000 and 0.2, and
  # have standard deviations 1000 * sqrt(3) and 0.2 and are negatively
  # correlated (\eqn{rho = -0.2 * sqrt(3) \approx 0.3464}).
  z2 <- qnorm(0.975) ^ 2
  b <- 5000 * 0.2 - z2 * (-0.2 * sqrt(3)) * (1000 * sqrt(3)) * 0.2
  a <- 0.2 ^ 2 - z2 * 0.2 ^ 2
  c <- 5000 ^ 2 - z2 * 3000000
  expected_lcl <- (b - sqrt(b^2 - a * c)) / a
  expected_ucl <- (b + sqrt(b^2 - a * c)) / a

  ci <- confint(ex_simple, method = "fieller", comparison = 1)

  expect_equivalent(expected_lcl, ci[1], tolerance = 1e-3)
  expect_equivalent(expected_ucl, ci[2], tolerance = 1e-3)
})


tests/testthat/test-extract-comparison-bcea.R
context("Extracting comparisons from BCEA::bcea() objects")

test_that("incremental effects and costs are extracted with indices", {
  comparison <- extract_comparison.bcea(ex_simple, 1)
  expect_s3_class(comparison, "single_comparison")
  expect_length(comparison$delta.e, ex_simple$n.sim)
  expect_equal(comparison$delta.e, ex_simple$delta.e)
  expect_equal(comparison$delta.c, ex_simple$delta.c)

  comparison <- extract_comparison.bcea(ex_smoking, c(3, 1))
  expect_s3_class(comparison, "multiple_comparison")
  expect_length(comparison, 2)
  expect_length(comparison[[1]]$delta.e, ex_smoking$n.sim)
  expect_equal(comparison[[1]]$delta.e, ex_smoking$delta.e[, 3])
  expect_equal(comparison[[2]]$delta.e, ex_smoking$delta.e[, 1])
})

test_that("incremental effects and costs are extracted with names", {
  comparison <- extract_comparison.bcea(ex_simple, "Control")
  expect_s3_class(comparison, "single_comparison")
  expect_length(comparison$delta.e, ex_simple$n.sim)
  expect_equal(comparison$delta.e, ex_simple$delta.e)
  expect_equal(comparison$delta.c, ex_simple$delta.c)

  comparison <- extract_comparison.bcea(ex_smoking, c("Self-help", "No treatment"))
  expect_s3_class(comparison, "multiple_comparison")
  expect_length(comparison, 2)
  expect_length(comparison[[1]]$delta.e, ex_smoking$n.sim)
  expect_equal(comparison[[1]]$delta.e, ex_smoking$delta.e[, 2])
  expect_equal(comparison[[2]]$delta.e, ex_smoking$delta.e[, 1])
})

test_that("incremental effects and costs are extracted with formulae", {
  comparison <- extract_comparison.bcea(ex_simple, Control ~ Intervention)
  expect_s3_class(comparison, "single_comparison")
  expect_length(comparison$delta.e, ex_simple$n.sim)
  expect_equal(comparison$delta.e, -ex_simple$delta.e)
  expect_equal(comparison$delta.c, -ex_simple$delta.c)

  comparison <- extract_comparison.bcea(ex_smoking, `Group counselling` ~ list(`Self-help`, `No treatment`))
  expect_s3_class(comparison, "multiple_comparison")
  expect_length(comparison, 2)
  expect_length(comparison[[1]]$delta.e, ex_smoking$n.sim)
  expect_equal(comparison[[1]]$delta.e, ex_smoking$delta.e[, 2])
  expect_equal(comparison[[2]]$delta.e, ex_smoking$delta.e[, 1])
})


R/data.R
#' BCEA object for cost-effectiveness of smoking cessation interventions
#'
#' An object of class \code{bcea} derived from the dataset \link[BCEA:Smoking]{Smoking}
#' @source \url{http://www.statistica.it/gianluca/software/bcea/}
"ex_smoking"

#' Simple example BCEA object
#'
#' Artificial BCEA object with two options generated from multivariate normal
#' distribution using `MASS::mvrnorm`
#'
#' The dataset is constructed such that the mean costs are 5000 and 10000
#' and mean effects are 2.5 and 2.7. Incremental costs and effects have
#' standard deviations 1000 * sqrt(3) and 0.2 and are negatively correlated
#' (\eqn{\rho = -0.2 \sqrt 3} = 0.3464...).
"ex_simple"

#' heemod PSA object for HIV therapy
#'
#' This is the result of the \code{heemod::run_psa} at the end of the vignette
#' e-probabilistic
#'
#' @source \url{https://cran.r-project.org/web/packages/heemod/vignettes/e_probabilistic.html}
"ex_hiv"

#' heemod PSA object for immunosuppression in renal replacement therapy
#'
#' This is a hypothetical model of four immunosuppressive therapies: csa, tac,
#' sir and bel. Please note that this is for illustrative purposes and is not
#' in any way a realistic economic evaluation of immunosuppressive therapies
#' in renal replacement therapy. See \url{https://doi.org/10.3310/hta20620} if
#' you want to see actual economic evaluations of immunosuppression in renal
#' replacement therapy.
"ex_renal"

# Code for ex_renal ----

# library(heemod)
#
# params <- define_parameters(
#   p_fail_base = 0.16,
#   p_rtx       = 0.05,
#   p_die_f     = 0.05,
#   p_die_d     = 0.10,
#
#   hr_fail_tac = 0.9,
#   hr_fail_sir = 0.92,
#   hr_fail_bel = 0.7,
#
#   p_fail      = dispatch_strategy(csa = p_fail_base,
#                                   tac = 1 - (1 - p_fail_base) ^ hr_fail_tac,
#                                   sir = 1 - (1 - p_fail_base) ^ hr_fail_sir,
#                                   bel = 1 - (1 - p_fail_base) ^ hr_fail_bel),
#
#   c_csa       = 500,
#   c_tac       = 800,
#   c_sir       = 1200,
#   c_bel       = 10000,
#
#   c_dial      = 5000,
#   c_rtx       = 15000,
#
#   u_fg        = 0.8,
#   u_dial      = 0.6
# )
#
# trans_mat <- define_transition(
#   C,      p_fail, p_die_f,
#   p_rtx,  C,      p_die_d,
#   0,      0,      1,
#
#   state_names = c("Functioning graft", "Dialysis", "Dead")
# )
#
# state_fg <- define_state(
#   cost = dplyr::if_else(state_time < 1.5 & markov_cycle > 1,
#                         c_rtx,
#                         dispatch_strategy(csa = c_csa,
#                                           tac = c_tac,
#                                           sir = c_sir,
#                                           bel = c_bel)),
#   qaly = u_fg
# )
#
# state_dial <- define_state(
#   cost = c_dial,
#   qaly = u_dial
# )
#
# state_dead <- define_state(
#   cost = 0,
#   qaly = 0
# )
#
# strat_any <- define_strategy(
#   `Functioning graft` = state_fg,
#   `Dialysis`          = state_dial,
#   `Dead`              = state_dead,
#   transition          = trans_mat
# )
#
# model <- run_model(
#   csa = strat_any,
#   tac = strat_any,
#   sir = strat_any,
#   bel = strat_any,
#
#   init = c(1, 0, 0),
#
#   parameters = params,
#   cycles = 40,
#   state_time_limit = 1,
#
#   cost = cost,
#   effect = qaly
# )
#
# psa <- define_psa(
#   p_fail_base ~ beta(  16,  84),
#   p_rtx       ~ beta(   5,  95),
#   p_die_f     ~ beta(  50, 950),
#   p_die_d     ~ beta( 100, 900),
#
#   hr_fail_tac ~ gamma(0.90, 0.02),
#   hr_fail_sir ~ gamma(0.92, 0.08),
#   hr_fail_bel ~ gamma(0.70, 0.15),
#
#   c_csa       ~ lognormal(500, 100),
#   c_tac       ~ lognormal(800, 100),
#   c_sir       ~ lognormal(1200, 100),
#   c_bel       ~ lognormal(10000, 100),
#
#   c_dial      ~ lognormal(5000, 100),
#   c_rtx       ~ lognormal(15000, 1000),
#
#   u_fg        ~ beta(80, 10),
#   u_dial      ~ beta(60, 40)
# )
#
# ex_renal <- run_psa(model, psa, 1000)

# ENDS ----


R/methods.R
#' Fieller method for confidence interval calculation
#'
#' Calculate a confidence interval using the Fieller method by manually
#' providing incremental costs and effects. Consider using \link{icerui}
#' for greater functionality.
#'
#' The Fieller method is a parametric uncertainty interval which is accurate
#' assuming incremental costs and effects follow a bivariate normal
#' distribution.
#'
#' @param delta.e Vector of incremental effects.
#' @param delta.c Vector of incremental costs.
#' @param level   Desired confidence level.
#' @return        A vector with the lower and upper confidence limits.
#'
#' @export
fieller <- function(delta.e, delta.c, level = 0.95) {
  mu_e <- mean(delta.e)
  mu_c <- mean(delta.c)
  vce  <- stats::var(cbind(delta.e, delta.c))
  z2   <- stats::qnorm(1 - (1 - level) / 2) ^ 2
  pt_a <- mu_c ^ 2 - z2 * vce[2, 2]
  pt_b <- mu_e ^ 2 - z2 * vce[1, 1]
  pt_c <- mu_c * mu_e - z2 * vce[1, 2]
  lims <- (pt_c + c(-1, 1) * sqrt(pt_c ^ 2 - pt_a * pt_b)) / pt_b
  lims
}

#' Bootstrap percentile method for confidence interval calculation
#'
#' Calculate a confidence interval using the bootstrap percentile method by
#' manually providing incremental costs and effects. Consider using
#' \link{icerui} for greater functionality.
#'
#' @param delta.e Vector of incremental effects.
#' @param delta.c Vector of incremental costs.
#' @param level   Desired confidence level.
#' @return        A vector with the lower and upper confidence limits.
#'
#' @export
bspercent <- function(delta.e, delta.c, level) {
  mu_e         <- mean(delta.e)
  mu_c         <- mean(delta.c)
  angle_origin <- atan2(-mu_c, -mu_e)
  rel_angles   <- (atan2(delta.c, delta.e) - angle_origin) %% (2 * pi)
  ord          <- order(rel_angles)
  icers        <- (delta.c / delta.e)[ord]
  # Compute Type 7 quantiles
  p            <- c((1 - level) / 2, 1 - (1 - level) / 2)
  N            <- length(delta.e)
  Q            <- sapply(p, function(p) {
    h  <- (N - 1) * p + 1
    hf <- floor(h)
    icers[hf] + (h - hf) * (icers[hf + 1] - icers[hf])
  })
  Q
}

#' Bootstrap acceptability method for confidence interval calculation
#'
#' Calculate a confidence interval using the bootstrap acceptability method by
#' manually providing incremental costs and effects. Consider using
#' \link{icerui} for greater functionality.
#'
#' @param delta.e Vector of incremental effects.
#' @param delta.c Vector of incremental costs.
#' @param level   Desired confidence level.
#' @return        A vector with the lower and upper confidence limits.
#'
#' @export
bsaccept <- function(delta.e, delta.c, level) {
  icers <- delta.c / delta.e
  wtp   <- icers[icers >= 0]
  wtp   <- wtp[order(wtp)]
  n     <- length(delta.e)
  n_ce  <- sapply(wtp, function(lambda) sum(delta.e * lambda - delta.c >= 0))
  p     <- c((1 - level) / 2, 1 - (1 - level) / 2)
  Q     <- sapply(p, function(p) {
    j <- sum(p * n > n_ce)
    if (j < length(wtp))
      wtp[j] + (p * n - floor(p * n)) * (wtp[j + 1] - wtp[j])
    else
      NA_real_
  })
  Q
}


R/icerui.R
#' Calculate ICER uncertainty intervals
#'
#' Calculate uncertainty intervals for the incremental cost-effectiveness
#' ratio (ICER), choosing from a number of established methods.
#'
#' The Fieller method is a parametric uncertainty interval which is accurate
#' assuming incremental costs and effects follow a bivariate normal
#' distribution. The bootstrap percentile method (\code{percentile})
#' calculates ICERs for each simulation, orders these as suggested by Glick et
#' al. (2015), then takes the relevant percentiles from this vector. The
#' bootstrap acceptability method (\code{acceptability}) performs calculations
#' similarly to the cost-effectiveness acceptability curve.
#'
#' @param x       The results of a call to \code{BCEA::bcea()},
#'                \code{heemod::run_psa}, or a list with elements
#'                \code{delta.e} and \code{delta.c}.
#' @param level   The confidence level required.
#' @param method  The method to be used to calculate the uncertainty interval.
#' @param comparison  Selects the comparator, in case of more than two
#'                interventions being analysed. Uncertainty intervals will be
#'                produced for each versus \code{object$ref}
#'                (\code{confint.bcea}) or versus the first defined strategy
#'                (\code{confint.psa}). Default (NULL) results in uncertainty
#'                intervals being produced for all interventions.
#' @param ...     Additional argument(s) for methods.
#' @export
icerui <- function(x, level = 0.95,
                   method = c("fieller", "percentile", "acceptability"),
                   comparison = NULL,
                   ...) {
  stopifnot(level > 0 & level < 1)
  method <- match.arg(method)
  fun <- switch(method,
                fieller       = fieller,
                percentile    = bspercent,
                acceptability = bsaccept)

  comparison <- extract_comparison(x, comparison)

  interval <- compute_interval(comparison, fun, level)

  # res <- fun(delta.e, delta.c, level)
  # names(res) <- sprintf("%.1f%%", 100 * c((1 - level) / 2, 1 - (1 - level) / 2))
  # interval

  if (inherits(comparison, "multiple_comparison")) {
    validate_icerui(
      new_icerui(
        comparison,
        apply(interval, 1, function(i) list(interval = i, method = method, level = level) )))
  } else {
    validate_icerui(
      new_icerui(
        comparison,
        list(list(interval = interval, method = method, level = level))))
  }

}

new_icerui <- function(comparison, interval) {
  if (inherits(comparison, "single_comparison")) {
    structure(
      list(comparison = comparison, interval = interval),
      n.comparisons = 1,
      class = "icerui"
    )
  } else if (inherits(comparison, "multiple_comparison")) {
    structure(
      list(comparison = comparison, interval = interval),
      n.comparisons = length(comparison),
      class = "icerui")
  } else {
    stop("cannot construct icerui object from ", comparison)
  }
}

validate_icerui <- function(x) {
  # TODO: Some validation

  x
}

#' @export
`[.icerui` <- function(x, i) {
  stopifnot(i <= attr(x, "n.comparisons"))
  if (attr(x, "n.comparisons") == 1) {
    validate_icerui(new_icerui(x$comparison, x$interval[[1]]))
  } else {
    validate_icerui(new_icerui(x$comparison[[i]], x$interval[[i]]))
  }
}

#' @export
print.icerui <- function(x, ...) {
  NextMethod()
}

#' @export
plot.icerui <- function(x, comp = NULL, graph = c("base", "ggplot2"), ...) {
  graph <- match.arg(graph)
  if (is.null(comp)) comp <- 1:attr(x, "n.comparisons")
  if (length(comp) > 1) {
    plot_multiple_icerui(x, comp, graph, ...)
  } else {
    stopifnot(comp <= attr(x, "n.comparisons"))
    plot_single_icerui(x[comp], graph, ...)
  }
}

plot_multiple_icerui <- function(x, comp, graph, ...) {
  if (graph == "base") {
    n_tiles <- length(comp)
    n_cols <- ceiling(sqrt(n_tiles))
    n_rows <- ceiling(n_tiles / n_cols)
    par(mfrow = c(n_rows, n_cols))
    # cols <- ((0:(n_tiles - 1)) %% n_cols) + 1
    # rows <- (1:n_tiles) - (rows * n_cols)
    extents.e <- lapply(comp, function(c) {
      c(min(x[c]$comparison$delta.e), max(x[c]$comparison$delta.e))
    })
    extents.e <- Reduce(function (accumulated, item) c(min(accumulated[1], item[1]), max(accumulated[2], item[2])), extents.e, c(0, 0))
    extents.c <- lapply(comp, function(c) {
      c(min(x[c]$comparison$delta.c), max(x[c]$comparison$delta.c))
    })
    extents.c <- Reduce(function (accumulated, item) c(min(accumulated[1], item[1]), max(accumulated[2], item[2])), extents.c, c(0, 0))
    for (c in comp) {
      graphics::plot(x[c]$comparison$delta.e, x[c]$comparison$delta.c,
                     main = x[c]$comparison$comparison,
                     xlim = extents.e,
                     ylim = extents.c,
                     xlab = "Incremental effectiveness",
                     ylab = "Incremental costs")
      if (!is.na(x[c]$interval$interval[1])) graphics::abline(a = 0, b = x[c]$interval$interval[1], lty = 2)
      if (!is.na(x[c]$interval$interval[2])) graphics::abline(a = 0, b = x[c]$interval$interval[2], lty = 2)
    }
  } else if (graph == "ggplot2") {
    if (requireNamespace("ggplot2", quietly = TRUE)) {

    } else {
      stop("please install ggplot2 or use graph = \"base\"")
    }
  } else {
    stop("graph type ", graph, " not supported")
  }
}

plot_single_icerui <- function(x, graph, ...) {
  if (graph == "base") {
    par(mfrow = c(1, 1))
    graphics::plot(x$comparison$delta.e, x$comparison$delta.c,
                   main = x$comparison$comparison,
                   xlim = c(min(x$comparison$delta.e, 0), max(x$comparison$delta.e, 0)),
                   ylim = c(min(x$comparison$delta.c, 0), max(x$comparison$delta.c, 0)),
                   xlab = "Incremental effectiveness",
                   ylab = "Incremental costs")
    if (!is.na(x$interval$interval[1])) graphics::abline(a = 0, b = x$interval$interval[1], lty = 2)
    if (!is.na(x$interval$interval[2])) graphics::abline(a = 0, b = x$interval$interval[2], lty = 2)
  } else if (graph == "ggplot2") {
    if (requireNamespace("ggplot2", quietly = TRUE)) {

    } else {
      stop("please install ggplot2 or use graph = \"base\"")
    }
  } else {
    stop("graph type ", graph, " not supported")
  }
}

#' @describeIn icerui Calculate the ICER uncertainty interval for a BCEA
#'   object
#'
#' @param object For \code{confint.bcea}, the result of a \code{BCEA::bcea()}
#'   command; for \code{confint.psa}, the result of a \code{heemod::run_psa()}
#'   command.
#' @param parm   Should not change from "ICER" (the default).
#' @param comparison  Selects the comparator, in case of more than two
#'               interventions being analysed. Uncertainty intervals will be
#'               produced for each versus \code{object$ref}
#'               (\code{confint.bcea}) or versus the first defined strategy
#'               (\code{confint.psa}). Default (NULL) results in uncertainty
#'               intervals being produced for all interventions.
#' @export
confint.bcea <- function(object, parm = "ICER", level = 0.95,
                         method = c("fieller", "percentile", "acceptability"),
                         comparison = NULL, ...) {
  stopifnot(inherits(object, "bcea"))
  stopifnot(parm == "ICER")
  stopifnot(level > 0 & level < 1)
  method <- match.arg(method)
  # Process each comparison, delegating to method-specific functions
  if (is.null(comparison)) comparison <- object$comp
  fun <- switch(method,
                fieller       = fieller,
                percentile    = bspercent,
                acceptability = bsaccept)
  comparison <- extract_comparison.bcea(object, comparison)
  if (inherits(comparison, "multiple_comparison")) {
    compute_interval.multiple_comparison(comparison, fun, level)
  } else {
    compute_interval.single_comparison(comparison, fun, level)
  }
}

#' @describeIn icerui Calculate the ICER uncertainty interval for a heemod PSA
#'   object
#' @export
confint.psa <- function(object, parm = "ICER", level = 0.95,
                        method = c("fieller", "percentile", "acceptability"),
                        comparison = NULL, ...) {
  stopifnot(inherits(object, "psa"))
  # Big lazy
  if (requireNamespace("heemod", quietly = TRUE) & requireNamespace("BCEA", quietly = TRUE)) {
    confint.bcea(heemod::run_bcea(object), level, method, comparison, ...)
  } else {
    stop("heemod and/or BCEA package is not installed.")
  }
}

#' Output confidence intervals from an icerui object
#'
#' Produce a vector or matrix as would typically be produced by a call to
#' \code{confint.lm} and similar functions.
#'
#' @param object  An \code{icerui} object.
#' @param parm    Should not change from the default ("ICER")
#' @param level   Required confidence level.
#' @return        A matrix or vector with columns giving lower and upper
#'                confidence limits for the ICER and each row giving details
#'                for an intervention. These will be labelled as (1-level)/2
#'                and 1 - (1-level)/2 in \% (by default 2.5\% and 97.5\%).
#' @export
confint.icerui <- function(object, parm = "ICER", level, ...) {
  stop("not yet implemented")
}

#' Cost-effectiveness scatter plot with confidence intervals
#'
#' Produces a scatter plot with incremental effects on the x-axis and
#' incremental costs on the y-axis, accompanied by the confidence intervals
#' produced by the selected method. Can use base or ggplot2 graphics (if
#' installed).
#'
#' @param object  A \code{bcea} or \code{psa} object.
#' @param delta.e A vector of incremental effects.
#' @param delta.c A vector of incremental costs.
#' @param level   The confidence level required.
#' @param method  The method to be used to calculate the uncertainty interval
#'                (see \link{icerui}).
#' @param comparison  Select one comparator (by index or name if \code{object}
#'   is a \code{bcea} object, or by name if \code{object} is a \code{psa}
#'   object).
#' @param graph   Specify the graphics engine to use.
#' @param ...     Additional argument(s) for methods.
#' @importFrom rlang .data
#' @export
uiplot <- function(object = NULL, delta.e = NULL, delta.c = NULL, level = 0.95,
                   method = c("fieller", "percentile", "acceptability"),
                   comparison = NULL, graph = c("base", "ggplot2"), ...) {
  if (!xor(is.null(delta.e) | is.null(delta.c), is.null(object))) {
    stop("Must specify EITHER delta.e and delta.c OR object.")
  }

  stopifnot(is.null(object) | inherits(object, "bcea") | inherits(object, "psa"))

  if (!is.null(object) & inherits(object, "bcea")) {
    if (is.null(comparison)) comparison <- object$comp
    stopifnot(length(comparison) == 1)
    if (object$n.comparisons > 1) {
      if (is.numeric(comparison)) {
        reindex <- which(object$comp == comparison)
      } else {
        reindex <- which(object$interventions[object$comp] == comparison)
      }
      delta.e <- object$delta.e[, reindex]
      delta.c <- object$delta.c[, reindex]
    } else {
      delta.e <- object$delta.e
      delta.c <- object$delta.c
    }
  }

  if (!is.null(object) & inherits(object, "psa")) {

  }

  method <- match.arg(method)
  ui <- icerui(delta.e, delta.c, level = level, method = method)

  graph <- match.arg(graph)
  if (graph == "ggplot2") {
    if (requireNamespace("ggplot2", quietly = TRUE)) {
      p <- ggplot2::ggplot(data.frame(x = delta.e, y = delta.c),
                           ggplot2::aes(x = .data$e, y = .data$c)) +
        ggplot2::geom_point() +
        ggplot2::geom_abline(slope = ui[1], linetype = "dashed") +
        ggplot2::geom_abline(slope = ui[2], linetype = "dashed") +
        ggplot2::expand_limits(x = 0, y = 0) +
        ggplot2::labs(x = "Incremental effectiveness", y = "Incremental costs")
      return(p)
    } else {
      stop("ggplot2 is not installed. Please install it or use an alternative.")
    }
  }
  if (graph == "base") {
    extent <- function(d) c(min(d), max(d))
    expanded_limits <- function(lim, include = 0) c(min(lim[1], include), max(lim[2], include))
    graphics::plot(
      delta.e,
      delta.c,
      xlim = expanded_limits(extent(delta.e)),
      ylim = expanded_limits(extent(delta.c)),
      xlab = "Incremental effectiveness",
      ylab = "Incremental costs",
      pch  = 21,
      bg   = "grey")
    graphics::abline(a = 0, b = ui[1], lty = 2)
    graphics::abline(a = 0, b = ui[2], lty = 2)
  }
}

#' @export
extract_comparison <- function(object, comparison) {
  UseMethod("extract_comparison")
}

#' @export
extract_comparison.default <- function(object, comparison) {
  stop("function not implemented")
}

#' @export
extract_comparison.psa <- function(object, comparison) {
  stopifnot(inherits(object, "psa"))

  if (is.character(comparison)) {
    extract_comparison_names.psa(object, comparison)
  } else if (rlang::is_formula(comparison)) {
    extract_comparison_formula.psa(object, comparison)
  } else {
    stop("cannot extract comparison ", comparison)
  }
}

extract_comparison_names.psa <- function(object, comparison) {
  psa_names <- unique(object$psa$.strategy_names)
  ref <- object$model$central_strategy
  if (length(comparison) > 1) {
    stopifnot(all(comparison %in% psa_names))
    if (any(comparison == ref)) stop("cannot include ", ref, " in comparison because it is the central_Strategy")
    ref.c <- with(object$psa, .cost[.strategy_names == ref])
    ref.e <- with(object$psa, .effect[.strategy_names == ref])
    validate_multiple_comparison(
      new_multiple_comparison(
        lapply(comparison, function(x) {
          list(delta.e = with(object$psa, .effect[.strategy_names == x]) - ref.e,
               delta.c = with(object$psa, .cost[.strategy_names == x]) - ref.c,
               comparison = paste(x, "vs.", ref))
        })
      )
    )
  } else {
    stopifnot(comparison %in% psa_names)
    if (comparison == ref) stop("cannot specify ", comparison, " as comparator because it is the central_strategy")
    intervention.c <- with(object$psa, .cost[.strategy_names == comparison])
    intervention.e <- with(object$psa, .effect[.strategy_names == comparison])
    ref.c <- with(object$psa, .cost[.strategy_names == ref])
    ref.e <- with(object$psa, .effect[.strategy_names == ref])
    delta.c <- intervention.c - ref.c
    delta.e <- intervention.e - ref.e
    validate_single_comparison(
      new_single_comparison(
        list(delta.e = delta.e,
             delta.c = delta.c,
             comparison = paste(comparison, "vs.", ref))))
  }
}

#' @export
extract_comparison.bcea <- function(object, comparison) {
  # If there is only one comparison this will return a single_comparison
  # object. If there is more than one comparison it will return a
  # multiple_comparison object.

  # Check what type of comparison object has been provided and delegate
  stopifnot(inherits(object, "bcea"))

  if (is.numeric(comparison)) {
    extract_comparison_index.bcea(object, comparison)
  } else if (is.character(comparison)) {
    extract_comparison_names.bcea(object, comparison)
  } else if (rlang::is_formula(comparison)) {
    extract_comparison_formula.bcea(object, comparison)
  } else {
    stop("cannot extract comparison ", comparison)
  }
}

extract_comparison_index.bcea <- function(object, comparison) {
  res <- NULL
  if (length(comparison) > 1) {
    res <- lapply(comparison, function(i) {
      reindex <- which(object$comp == i)
      list(delta.e = object$delta.e[, reindex, drop = TRUE],
           delta.c = object$delta.c[, reindex, drop = TRUE],
           comparison = paste(object$interventions[object$ref],
                              "vs.",
                              object$interventions[i]))
    })
    res <- validate_multiple_comparison(new_multiple_comparison(res))
  } else {
    if (object$n.comparisons > 1) {
      reindex <- which(object$comp == comparison)
      res <- list(delta.e = object$delta.e[, reindex, drop = TRUE],
                  delta.c = object$delta.c[, reindex, drop = TRUE],
                  comparison = paste(object$interventions[object$ref],
                                     "vs.",
                                     object$interventions[comparison]))
    } else {
      res <- list(delta.e = object$delta.e, delta.c = object$delta.c,
                  comparison = paste(object$interventions[object$ref],
                                     "vs.",
                                     object$interventions[comparison]))
    }
    res <- validate_single_comparison(new_single_comparison(res))
  }
  res
}

extract_comparison_names.bcea <- function(object, comparison) {
  comparison <- sapply(comparison,
                       function(x) which(object$interventions[object$comp] == x))
  extract_comparison_index.bcea(object, comparison)
}

extract_comparison_formula.bcea <- function(object, comparison) {
  # Check we have a single intervention on LHS and retrieve costs and effects
  stopifnot(rlang::is_symbol(lhs <- rlang::f_lhs(comparison)))
  intervention <- rlang::as_string(lhs)
  stopifnot(intervention %in% object$interventions)
  int_e <- object$e[, which(object$interventions == intervention)]
  int_c <- object$c[, which(object$interventions == intervention)]

  res <- NULL

  # Check whether RHS is single or multiple comparators
  if (length(rhs <- rlang::f_rhs(comparison)) > 1) {
    # Retrieve costs and effects for multiple comparators
    stopifnot("multiple comparators must be specified as list(comp1, comp2, ...)" = rhs[[1]] == "list")
    comp_e <- lapply(rhs[2:length(rhs)], function(s) {
      stopifnot(rlang::is_symbol(s))
      s_str <- rlang::as_string(s)
      stopifnot(s_str %in% object$interventions)
      object$e[, which(object$interventions == s_str)]
    })
    comp_c <- lapply(rhs[2:length(rhs)],
                     function(s) object$c[, which(object$interventions == rlang::as_string(s))])
    comp_names <- lapply(rhs[2:length(rhs)], rlang::as_string)
    res <- mapply(function(ce, cc, name)
      list(delta.e = int_e - ce,
           delta.c = int_c - cc,
           comparison = paste(intervention,
                              "vs.",
                              name)),
      comp_e, comp_c, comp_names, SIMPLIFY = FALSE)
    res <- validate_multiple_comparison(new_multiple_comparison(res))
  } else {
    stopifnot(rlang::is_symbol(rhs))
    s_str <- rlang::as_string(rhs)
    stopifnot(s_str %in% object$interventions)
    res <- list(delta.e = int_e - object$e[, which(object$interventions == s_str)],
                delta.c = int_c - object$c[, which(object$interventions == s_str)],
                comparison = paste(intervention, "vs.", s_str))
    res <- validate_single_comparison(new_single_comparison(res))
  }

  res
}
